### PR TITLE
Homework3: Aleksandr Strokov

### PR DESCRIPTION
**Pull request by alstrokov**

### CODE REVIEW

File: src/task1.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+"""List practice"""
+
+# 1
+list1 = [ltr1 + ltr2 for ltr1 in ['a', 'b'] for ltr2 in ['b', 'c', 'd']]
```
Comment (belovmd) [inline src/task1.py]: 'ab'?

File: src/task1.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+print(list1)
+
+# 2
+list2 = list1[::2]
+print(list2)
+
+# 3
+list3 = [str(x) + 'a' for x in range(1, 5)]
```
Comment (belovmd) [inline src/task1.py]: do we need range here?

File: src/task1.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+print(list3)
+
+# 4
+print(list3.pop(1))
+print(list3)
+
+# 5
+list4 = list3.copy()
+list4.append('2a')
```
Comment (belovmd) [inline src/task1.py]: it would be in the end

File: src/task2.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
+"""Tuple practice"""
+# 1
+lst = ['a', 'b', 'c']
+tpl = tuple(lst)
+
+# 2
+tpl = ('a', 'b', 'c')
+lst = list(tpl)
+
+print("{}: {}".format(type(lst), lst))
```
Comment (belovmd) [inline src/task2.py]: why do we have this prints but print nothing above? it seems it should be similar

File: src/task3.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
+"""Dictionary practice"""
+
+
+# 1
+def generate_numbers(num=20):
+    dct = dict()
```
Comment (belovmd) [inline src/task3.py]: can we use dict comprehension here?

File: src/task3.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+print(generate_numbers(3))
+print(generate_numbers())
+
+
+# 2
+def count_characters(count_me_string):
+    dct = dict()
+
+    for char in count_me_string:
+        if char in dct.keys():
```
Comment (belovmd) [inline src/task3.py]: we have more native way here. please take a look at dict methods

File: src/task3.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+            dct[char] += 1
+        else:
+            dct[char] = 1
+
+    return dct
+
+
+print(count_characters("abcdefgabc"))
```
Comment (belovmd) [inline src/task3.py]: redundant print

File: src/task3.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+def generate_numbers(num=20):
+    dct = dict()
+
+    for el in range(num + 1):
+        dct[el] = el**2
+
+    return dct
+
+
+print(generate_numbers(3))
```
Comment (belovmd) [inline src/task3.py]: redundant prints

**Review by belovmd — COMMENTED** [review]
Please add tasks description to files
Please address next issues:

File: src/task1.py (hunk 1 above)
Comment (alstrokov) [reply src/task1.py]: Yes, 'ab' looks much more accurate

File: src/task1.py (hunk 2 above)
Comment (alstrokov) [reply src/task1.py]: Is plain string acceptable solution here?

File: src/task1.py (hunk 3 above)
Comment (alstrokov) [reply src/task1.py]: Yes, in task description it wasn't explicitly declared that `2a` should be added to its original position

**Comment by alstrokov** [comment]
Homework updated and ready for code review

File: src/task3.py [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+# 1
+def generate_numbers(num=20):
+    return {el: el**2 for el in range(num + 1)}
+
+
+# 2
+def count_characters(count_me_string):
+    return {
+        char: count_me_string.count(char)
```
Comment (belovmd) [inline src/task3.py:26]: now it's O(n**2)
dct[char] = dct.get(char, 0) +1

**Review by belovmd — COMMENTED** [review]
7.5

**Pull request closed without merging**